pkg/capabilities/triggers/mercury_trigger.go

```diff
@@ -225,19 +171,76 @@ func (o *MercuryTriggerService) GetTriggerID(req capabilities.CapabilityRequest,
 	return "", fmt.Errorf("triggerId not found in inputs")
 }
 
-func GenerateTriggerEventID(reports []mercury.FeedReport) string {
-	// Let's hash all the feedIds and timestamps together
-	sort.Slice(reports, func(i, j int) bool {
-		if reports[i].FeedID == reports[j].FeedID {
-			return reports[i].ObservationTimestamp < reports[j].ObservationTimestamp
+func (o *MercuryTriggerService) loop() {
+	defer o.wg.Done()
+	now := time.Now().UnixMilli()
+	nextWait := o.tickerResolutionMs - now%o.tickerResolutionMs
+
+	for {
+		select {
+		case <-o.stopCh:
+			return
+		case <-time.After(time.Duration(nextWait) * time.Millisecond):
+			startTs := time.Now().UnixMilli()
+			// find closest timestamp that is a multiple of o.tickerResolutionMs
+			aligned := (startTs + o.tickerResolutionMs/2) / o.tickerResolutionMs * o.tickerResolutionMs
+			o.process(aligned)
+			currTs := time.Now().UnixMilli()
+			desiredNext := aligned + o.tickerResolutionMs
+			nextWait = desiredNext - currTs
+			if nextWait <= 0 {
+				o.lggr.Errorw("processing took longer than ticker resolution", "duration", currTs-startTs)
+				nextWait = 0
+			}
+		}
+	}
+}
+
+func (o *MercuryTriggerService) process(timestamp int64) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	for _, sub := range o.subscribers {
+		if timestamp%int64(sub.config.MaxFrequencyMs) == 0 {
+			reportList := make([]mercury.FeedReport, 0)
+			for _, feedID := range sub.config.FeedIds {
+				if latest, ok := o.latestReports[mercury.FeedID(feedID)]; ok {
+					reportList = append(reportList, latest)
+				}
+			}
+
+			val, err := mercury.Codec{}.Wrap(reportList)
+			if err != nil {
+				o.lggr.Errorw("Error encoding reportList", "err", err)
+				continue
+			}
+
+			timestampStr := strconv.FormatInt(timestamp, 10)
+			triggerEvent := capabilities.TriggerEvent{
+				TriggerType: "mercury",
+				ID:          timestampStr, // use timestamp as EventID
+				Timestamp:   timestampStr,
+				Payload:     val,
+			}
+
+			eventVal, err := values.Wrap(triggerEvent)
+			if err != nil {
+				o.lggr.Errorw("Error wrapping triggerEvent", "err", err)
+				continue
+			}
+
+			// Create a new CapabilityResponse with the MercuryTriggerEvent
+			capabilityResponse := capabilities.CapabilityResponse{
+				Value: eventVal,
+			}
+
+			o.lggr.Debugw("ProcessReport pushing event", "nReports", len(reportList), "eventID", triggerEvent.ID)
+			select {
+			case sub.ch <- capabilityResponse:
+			default:
```

Comment:
Don't we want subscribers to always get the latest feed? I'd imagine that we want to keep a small buffer and aggresively drop the head elements

Reply:
That shouldn't be a concern of the trigger. It's up to the consumer to handle it, if they want to. In fact, in the remote setting, we already have a ring buffer of messages.